ui/src/AppCharting/chart-data-extractor.ts

```diff
@@ -0,0 +1,220 @@
+ * @link C3 Types https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/c3/index.d.ts
+ */
+export type ChartingColumns = Array<ChartingArray>;
+
+/**
+ * Chart Data Strucutre for C3 Charts
+ * @link C3 Types https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/c3/index.d.ts
+ */
+export type ChartFactoryData = Data;
+
+export type ChartFactoryCallback = (
+  benchmarkMap: Benchmarks,
+  timestamps: TimestampViewModel[],
+  opts: FormConfiguration
+) => Promise<Data>;
+
+const FUNCTION_MAP: Record<ChartType, ChartFactoryCallback> = {
+  [BY_TIMESTAMP]: getChartDataByTimestamp,
+  [BY_SIZE]: getChartDataBySize
+};
+
+export function classNameFromBenchmarkKey(key: BenchmarkIdentifier): string {
+  const [date] = key.split('_');
+  const split = date.split('+');
+  if (!split.length) return '';
+  const first: string = split[0];
+
+  return first.replaceAll('T', '_').replaceAll(':', '-');
+}
+
+export function classNameToDateTime(text: string): string {
+  const [date, time] = text.split('_');
+  const newTime = time.replaceAll('-', ':');
+  return date + ' ' + newTime;
+}
+
+export function chartDataFactory(type: ChartType): ChartFactoryCallback {
+  const fn = FUNCTION_MAP[type];
+  if (!fn) throw new Error(`Chart data for ${type} is not currently supported`);
+
+  return async (
+    benchmarkMap: Benchmarks,
+    timestamps: TimestampViewModel[],
+    opts: FormConfiguration
+  ): Promise<Data> => {
+    const data = await fn(benchmarkMap, timestamps, opts);
+    return mutating_assignNames(data);
+  };
+}
+
+export function getChartDataBySize(
+  benchmarkMap: Benchmarks,
+  timestamps: TimestampViewModel[],
+  opts: FormConfiguration
+): Promise<Data> {
+  const {scenario, serverCount, plotType, statName} = opts;
+  const isFan = isFanScenario(scenario);
+  const data: Data = {columns: [], x: 'x', names: {}};
+
+  if (!scenario || !statName || !plotType) return Promise.resolve(data);
+
+  const sizes = getSizeKeys(benchmarkMap, {scenario, serverCount});
+  const arr: ChartingArray = ['x', ...sizes];
+  const columns: ChartingColumns = [arr];
+
+  for (const timestamp of timestamps) {
+    const column: ChartingArray = [classNameFromBenchmarkKey(timestamp.key)];
+
+    for (const size of sizes) {
+      let value: number = MISSING_VALUE;
+      const sizeKey = isFan ? `${size}_${serverCount}` : size;
+      const benchmark = benchmarkMap[timestamp.key];
+      const scenarioSizeRecords: ScenarioSizeRecords | boolean =
+        benchmark && benchmark[scenario];
+      if (scenarioSizeRecords) {
+        const plotTypeRecords: PlotTypeRecords = scenarioSizeRecords[sizeKey];
+        if (plotTypeRecords) {
+          const plotStatistic: PlotStatistic = plotTypeRecords[plotType];
+          if (plotStatistic && plotStatistic[statName])
+            value = plotStatistic[statName];
+        }
+      }
+
+      column.push(value);
+    }
+    columns.push(column);
+  }
+
+  return Promise.resolve({...data, columns});
```

Comment:
Why does this function and the next return a Promise? Is it because the caller might call a different function instead that really does something asynchronous and doing this allows the caller to assume that a Promise is always returned?

Reply:
Yeah.. I wanted to make all of these chart functions have a return type
and in the event that for an unforeseen reason some async action needed to be executed within any of the functions it wouldn't trip up.  

This specifically, was to make TS stop complaining about no awaited actions within an async function.

Reply:
If I remember correctly, JavaScript lets you use `await` on functions that do not return a Promise and it works fine. So I think it work regardless of whether the function called after `await` returns a Promise. If that is true, you can remove `async` from your function and return a plain value rather than the result of `Promise.resolve(value)`.

Reply:
Oh yeah... you're absolutely right. 

```
export async function getChartDataByTimestamp
```

did sort out my TS warning

Fixed 8b81a4058af56244e5fe62cf8619ca3927cf38c6